src/libserver/redis_pool.cxx

```diff
@@ -487,9 +558,36 @@ auto redis_pool_elt::new_connection() -> redisAsyncContext *
 							conn->ctx->errstr, ip.c_str(), port, nctx);
 
 			if (nctx) {
-				active.emplace_front(std::make_unique<redis_pool_connection>(pool, this,
-																			 db.c_str(), username.c_str(), password.c_str(), nctx));
-				active.front()->elt_pos = active.begin();
+				/* If TLS is configured for this element, initiate it now */
+				if (use_tls && !is_unix) {
+					if (ensure_ssl_inited()) {
+						redisSSLContextError ssl_err = REDIS_SSL_CTX_NONE;
+						redisSSLOptions opts{};
+						opts.cacert_filename = ca_file.empty() ? nullptr : ca_file.c_str();
+						opts.capath = ca_dir.empty() ? nullptr : ca_dir.c_str();
+						opts.cert_filename = cert_file.empty() ? nullptr : cert_file.c_str();
+						opts.private_key_filename = key_file.empty() ? nullptr : key_file.c_str();
+						opts.server_name = sni.empty() ? nullptr : sni.c_str();
+						opts.verify_mode = no_ssl_verify ? REDIS_SSL_VERIFY_NONE : REDIS_SSL_VERIFY_PEER;
+
+						auto *ssl_ctx = redisCreateSSLContextWithOptions(&opts, &ssl_err);
+						if (!ssl_ctx || redisInitiateSSLWithContext(&nctx->c, ssl_ctx) != REDIS_OK) {
+							msg_err("cannot start TLS for redis %s:%d: %s", ip.c_str(), port, nctx->errstr);
+							if (ssl_ctx) redisFreeSSLContext(ssl_ctx);
+							redisAsyncFree(nctx);
+							nctx = nullptr;
+						}
+						else {
+							redisFreeSSLContext(ssl_ctx);
+						}
+					}
+				}
+
+				if (nctx) {
+					active.emplace_front(std::make_unique<redis_pool_connection>(pool, this,
+																		 db.c_str(), username.c_str(), password.c_str(), nctx));
+					active.front()->elt_pos = active.begin();
+				}
```

Comment:
We repeat this code part 2 times, it's good to have it as a separate functor/lambda to avoid cut&paste.

Reply:
Pulled it into a helper